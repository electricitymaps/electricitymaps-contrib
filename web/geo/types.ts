import { Feature, FeatureCollection, MultiPolygon, Polygon } from '@turf/turf';

import { GEO_CONFIG } from './generateWorld';

export type GeoConfig = typeof GEO_CONFIG;

export interface ZoneConfig {
  subZoneNames?: string[];
  bounding_box?: number[][];
  contributors?: string[];
  disclaimer?: string;
  estimation_method?: string;
  generation_only?: boolean;
  parsers?: {
    consumption?: string;
    consumptionForecast?: string;
    generationForecast?: string;
    price?: string;
    production?: string;
    productionPerModeForecast?: string;
    productionPerUnit?: string;
  };
  [key: string]:
    | undefined
    | string[]
    | number[][]
    | string
    | boolean
    | { [key: string]: string };
}

export interface OptimizedZoneConfig {
  contributors?: number[];
  disclaimer?: string;
  estimation_method?: string;
  parsers: boolean;
  subZoneNames?: string[];
  generation_only?: boolean;
  timezone?: string;
}

export interface ZonesConfig {
  [key: string]: ZoneConfig;
}

export interface OptimizedZonesConfig {
  [key: string]: OptimizedZoneConfig;
}

export interface CombinedZonesConfig {
  contributors: string[];
  zones: OptimizedZonesConfig;
}

export interface ExchangeConfig {
  lonlat: [number, number];
  rotation: number;
  [key: string]: [number, number] | number;
  // The following properties are removed from the generated exchange config
  // and are either not used in the app or are provided by the backend instead.
  // capacity?: [number, number];
  // comment?: string;
  // _comment?: string;
  // parsers?: {
  //   exchange?: string;
  //   exchangeForecast?: string;
  // };
}

export interface ExchangesConfig {
  [key: string]: ExchangeConfig;
}

export declare type FeatureProperties = {
  zoneName: string;
  countryKey: string;
  isAggregatedView?: boolean;
  isHighestGranularity?: boolean;
  isCombined?: boolean;
};

export interface WorldFeatureCollection extends FeatureCollection {
  features: Feature<MultiPolygon | Polygon, FeatureProperties>[];
}
