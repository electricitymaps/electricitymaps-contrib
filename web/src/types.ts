import type {
  Feature,
  FeatureCollection,
  Geometry,
  MultiPolygon,
  Polygon,
} from '@turf/turf';
import { LineString, MultiLineString, Point } from 'geojson';
import { EstimationMethods, TimeRange } from 'utils/constants';

export type Maybe<T> = T | null | undefined;

export type ZoneKey = string;

export interface GridState {
  _disclaimer: string;
  createdAt: string;
  datetimes: {
    /** Object representing the grid state at a single point in time */
    [datetimeKey: string]: {
      /** Array of all exchanges */
      e: {
        [key: ZoneKey]: StateExchangeData;
      };
      /** Array of all zones */
      z: {
        [key: ZoneKey]: StateZoneData;
      };
    };
  };
}

export interface StateZoneData {
  /** Object representing all production values */
  p: {
    /** Carbon intensity */
    ci?: number | null;
    /** Fossil ratio */
    fr?: number | null;
    /** Renewable ratio */
    rr?: number | null;
  };
  /** Object representing all consumption values */
  c: {
    /** Carbon intensity */
    ci?: number | null;
    /** Fossil ratio */
    fr?: number | null;
    /** Renewable ratio */
    rr?: number | null;
  };
  /** Estimation method */
  em?: EstimationMethods | null;
  /** Estimation percentage */
  ep?: number | null;
  /** Represents if the zone has a outage message or not */
  o?: boolean | null;
}

export interface StateExchangeData {
  /** The carbon intensity of the exchange */
  ci: number;
  /** The net flow of the exchange */
  f: number;
}

export interface ExchangeArrowData {
  co2intensity: number;
  netFlow: number;
  rotation: number;
  lonlat: [number, number];
  key: string;
}

export interface ZoneOverviewForTimePeriod {
  [dateTimeKey: string]: ZoneOverview;
}

export interface ZoneOverview {
  zoneKey: string;
  co2intensity?: number;
  co2intensityProduction?: number;
  stateDatetime: string;
  fossilFuelRatio: number;
  renewableRatio: number;
  estimationMethod?: EstimationMethods;
  estimatedPercentage?: number;
}

export type GenerationType =
  | 'biomass'
  | 'coal'
  | 'gas'
  | 'hydro'
  | 'nuclear'
  | 'oil'
  | 'solar'
  | 'unknown'
  | 'geothermal'
  | 'wind';

export type ElectricityStorageType = 'hydro storage' | 'battery storage';
export type ElectricityStorageKeyType = 'battery' | 'hydro';

export type ElectricityModeType = GenerationType | ElectricityStorageType;

export type Exchange = { [key: string]: number };

export interface ZoneDetail extends ZoneOverview {
  _isFinestGranularity: boolean;
  estimatedPercentage?: number;
  measuredPercentage?: number;
  completenessPercentage?: number;
  // Capacity is only available on hourly details
  capacity?: { [key in ElectricityModeType]: number | null };
  capacitySources?: { [key in ElectricityModeType]: string[] | null };
  dischargeCo2Intensities: { [key in ElectricityStorageKeyType]: number };
  dischargeCo2IntensitySources: { [key in ElectricityStorageKeyType]: string };
  exchange: Exchange;
  exchangeCapacities?: {
    [key: ZoneKey]: number[];
  };
  exchangeCo2Intensities: Exchange;
  fossilFuelRatio: number;
  fossilFuelRatioProduction: number;
  isValid: boolean;
  maxCapacity: number;
  maxDischarge: number;
  maxExport: number;
  maxExportCapacity: number;
  maxImport: number;
  maxImportCapacity: number;
  maxProduction: number;
  maxStorage: number;
  maxStorageCapacity: number;
  price?: {
    value: number;
    currency: string;
    disabledReason?: string;
  };
  production: { [key in GenerationType]: Maybe<number> };
  productionCo2Intensities: { [key in GenerationType]: number };
  productionCo2IntensitySources: { [key in GenerationType]: string };
  renewableRatio: number;
  renewableRatioProduction: number;
  source: string[];
  storage: { [key in ElectricityStorageKeyType]: Maybe<number> };
  totalCo2Consumption: number;
  totalCo2Discharge: number | null;
  totalCo2Export: number | null;
  totalCo2Import: number | null;
  totalCo2NetExchange: number | null;
  totalCo2Production: number;
  totalCo2Storage: number | null;
  totalConsumption: number;
  totalDischarge: number | null;
  totalExport: number | null;
  totalImport: number | null;
  totalProduction: number | null;
  totalStorage: number | null;
}

export interface ZoneDetails {
  hasData: boolean;
  futurePrice: FuturePriceData;
  //TODO Remove from backend, most likely unused
  stateAggregation: 'daily' | 'hourly' | 'monthly' | 'yearly';
  zoneStates: {
    [key: string]: ZoneDetail;
  };
  zoneMessage?: ZoneMessage;
}

export interface ZoneMessage {
  message: string;
  issue?: string;
}

export interface GeometryProperties {
  center: [number, number];
  color: string;
  countryKey: string;
  isAggregatedView: boolean;
  isHighestGranularity: boolean;
  zoneId: string;
  zoneName: string;
}
export interface StateGeometryProperties {
  center?: [number, number];
  stateName?: string;
  stateId?: string;
}

export interface MapGeometries extends FeatureCollection<Geometry> {
  features: Array<MapGeometry>;
}

export interface StatesGeometries extends FeatureCollection<Geometry> {
  features: Array<StatesGeometry>;
}
export interface MapGeometry extends Feature<Polygon | MultiPolygon> {
  geometry: MultiPolygon | Polygon;
  Id?: number;
  properties: GeometryProperties;
}

export interface StatesGeometry extends Feature<LineString | MultiLineString | Point> {
  geometry: LineString | MultiLineString | Point;
  Id?: number;
  properties: StateGeometryProperties;
}

export interface MapTheme {
  co2Scale: {
    steps: number[];
    colors: string[];
  };
  oceanColor: string;
  strokeWidth: number;
  strokeColor: string;
  stateBorderColor: string;
  clickableFill: string;
  nonClickableFill: string;
}

export interface FuturePriceData {
  entryCount: number;
  priceData: {
    [key: string]: number;
  };
  currency: string;
  source: string;
  zoneKey: ZoneKey;
}

// Type for the URL parameters that determine app state
export type RouteParameters = {
  resolution?: string;
  zoneId?: string;
  urlTimeRange?: TimeRange;
  urlDatetime?: string;
};
