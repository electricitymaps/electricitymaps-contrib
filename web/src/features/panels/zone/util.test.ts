import { ZoneDetails } from 'types';
import { describe, expect, it } from 'vitest';

import {
  getContributors,
  getDisclaimer,
  getHasSubZones,
  getZoneDataStatus,
  isGenerationOnlyZone,
  showEstimationFeedbackCard,
  ZoneDataStatus,
} from './util';

describe('getContributors', () => {
  it('should return contributors for a zone', () => {
    const result = getContributors('FI');
    expect(result).toMatchSnapshot();
  });
});

describe('getDisclaimer', () => {
  it('should return disclaimer for a zone', () => {
    const result = getDisclaimer('NL');
    expect(result).toMatchSnapshot();
  });
});

describe('getHasSubZones', () => {
  it('should return null when the zoneId is falsy', () => {
    const result = getHasSubZones(undefined);
    expect(result).toEqual(null);
  });

  it('should return false when zone has no subzones', () => {
    const result = getHasSubZones('FI');
    expect(result).toEqual(false);
  });

  it('should return true when zone has subzones', () => {
    const result = getHasSubZones('SE');
    expect(result).toEqual(true);
  });
});

describe('getZoneDataStatus', () => {
  it('should return ZoneDataStatus.UNKNOWN when zoneDetails is falsy', () => {
    const result = getZoneDataStatus('FI', undefined as unknown as ZoneDetails);
    expect(result).toEqual(ZoneDataStatus.UNKNOWN);
  });

  it.todo('should return ZoneDataStatus.AVAILABLE when zoneDetails.hasData is true');

  it.todo(
    'should return ZoneDataStatus.NO_INFORMATION when there is no config for the zone'
  );

  it.todo(
    'should return ZoneDataStatus.NO_INFORMATION when there are no production parsers and no defined estimation method in the config'
  );

  it.todo('should return ZoneDataStatus.NO_REAL_TIME_DATA otherwise');
});

describe('isGenerationOnlyZone', () => {
  it('should return true when zone is generation only', () => {
    const result = isGenerationOnlyZone('US-CAR-YAD');
    expect(result).toEqual(true);
  });

  it('should return false when zone is not generation only', () => {
    const result = isGenerationOnlyZone('DK-DK1');
    expect(result).toEqual(false);
  });
});

describe('showEstimationFeedbackCard', () => {
  beforeEach(() => {
    localStorage.clear();
  });

  it('should return true when feedbackCard is currently not shown, card has not been seen before and it has been collapsed', () => {
    const result = showEstimationFeedbackCard(1, false, false, () => {});
    expect(result).toEqual(true);
  });

  it('should return true when isFeedbackCardVisibile is true', () => {
    const result = showEstimationFeedbackCard(1, true, true, () => {});
    expect(result).toEqual(true);
  });

  it('should return false when conditions are not met', () => {
    const result = showEstimationFeedbackCard(0, false, true, () => {});
    expect(result).toEqual(false);
  });

  it('should return false when conditions are not met', () => {
    const result = showEstimationFeedbackCard(2, false, true, () => {});
    expect(result).toEqual(false);
  });
});
