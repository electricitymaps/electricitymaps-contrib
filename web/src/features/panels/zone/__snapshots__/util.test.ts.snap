// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getContributors > should return contributors for a zone 1`] = `
[
  "corradio",
  "nessie2013",
  "gadakast",
  "pnook",
]
`;

exports[`getDisclaimer > should return disclaimer for a zone 1`] = `"Solar production is collected from NED.nl and can contain their own estimations."`;
